server/src-lib/Hasura/GraphQL/Transport/HTTP.hs

```diff
@@ -84,64 +120,112 @@ runGQ
+  (telemTimeTot_DT, (telemQueryType, telemTimeIO_DT, telemLocality, resp)) <- withElapsedTime $ do
     E.ExecutionCtx _ sqlGenCtx pgExecCtx {- planCache -} sc scVer httpManager enableAL <- ask
 
     -- run system authorization on the GraphQL API
     reqParsed <- E.checkGQLExecution userInfo (reqHeaders, ipAddress) enableAL sc reqUnparsed
                  >>= flip onLeft throwError
 
-    (telemCacheHit, execPlan) <- E.getResolvedExecPlan env logger pgExecCtx {- planCache -}
+    (_telemCacheHit, execPlan) <- E.getResolvedExecPlan env logger pgExecCtx {- planCache -}
                                  userInfo sqlGenCtx sc scVer queryType
                                  httpManager reqHeaders (reqUnparsed, reqParsed)
     case execPlan of
-      E.QueryExecutionPlan queryPlan asts ->
-        case queryPlan of
-          E.ExecStepDB txGenSql -> do
-            (telemTimeIO, telemQueryType, respHdrs, resp) <-
-              runQueryDB reqId (reqUnparsed,reqParsed) asts userInfo txGenSql
-            return (telemCacheHit, Telem.Local, (telemTimeIO, telemQueryType, HttpResponse resp respHdrs))
-          E.ExecStepRemote (rsi, opDef, _varValsM) ->
-            runRemoteGQ telemCacheHit rsi opDef
-          E.ExecStepRaw (name, json) -> do
-            (telemTimeIO, obj) <- withElapsedTime $
-              return $ encJFromJValue $ J.Object $ Map.singleton (G.unName name) json
-            return (telemCacheHit, Telem.Local, (telemTimeIO, Telem.Query, HttpResponse obj []))
-      E.MutationExecutionPlan mutationPlan ->
-        case mutationPlan of
-          E.ExecStepDB (tx, responseHeaders) -> do
-            (telemTimeIO, telemQueryType, resp) <- runMutationDB reqId reqUnparsed userInfo tx
-            return (telemCacheHit, Telem.Local, (telemTimeIO, telemQueryType, HttpResponse resp responseHeaders))
-          E.ExecStepRemote (rsi, opDef, _varValsM) ->
-            runRemoteGQ telemCacheHit rsi opDef
-          E.ExecStepRaw (name, json) -> do
-            (telemTimeIO, obj) <- withElapsedTime $
-              return $ encJFromJValue $ J.Object $ Map.singleton (G.unName name) json
-            return (telemCacheHit, Telem.Local, (telemTimeIO, Telem.Query, HttpResponse obj []))
+      E.QueryExecutionPlan queryPlans asts -> trace "Query" $ do
+        let cacheKey = QueryCacheKey reqParsed $ _uiRole userInfo
+        (responseHeaders, cachedValue) <- Tracing.interpTraceT id $ cacheLookup asts cacheKey
+        case cachedValue of
+          Just cachedResponseData ->
+            pure (Telem.Query, 0, Telem.Local, HttpResponse cachedResponseData responseHeaders)
+          Nothing -> do
+            conclusion <- runExceptT $ forWithKey queryPlans $ \fieldName -> \case
+              E.ExecStepDB txGenSql -> doQErr $ do
+                (telemTimeIO_DT, resp) <-
+                  runQueryDB reqId reqUnparsed userInfo fieldName
+                  (Just . Map.singleton (G.unsafeMkName fieldName) <$> txGenSql)
```

Comment:
Could `QueryExecutionPlan` be changed to include a `Name` already so that we don’t need to use `unsafeMkName`?

Reply:
Done.